pkg/descheduler/descheduler.go

```diff
@@ -0,0 +1,413 @@
+		return err
+	}
+	rs.Client = rsclient
+	rs.EventClient = eventClient
+
+	deschedulerPolicy, err := LoadPolicyConfig(rs.PolicyConfigFile, rs.Client, pluginregistry.PluginRegistry)
+	if err != nil {
+		return err
+	}
+	if deschedulerPolicy == nil {
+		return fmt.Errorf("deschedulerPolicy is nil")
+	}
+	deschedulerPolicyJson, err := json.MarshalIndent(deschedulerPolicy, "", "  ")
+	if err != nil {
+		return fmt.Errorf("faild to marshal deschedulerPolicy: %v", err)
+	}
+	klog.V(5).InfoS("Successfully load descheduler policy", "deschedulerPolicy", string(deschedulerPolicyJson))
+
+	// Add k8s compatibility warnings to logs
+	versionCompatibilityCheck(rs)
+
+	evictionPolicyGroupVersion, err := eutils.SupportEviction(rs.Client)
+	if err != nil || len(evictionPolicyGroupVersion) == 0 {
+		return err
+	}
+
+	runFn := func() error {
+		return RunDeschedulerStrategies(ctx, rs, deschedulerPolicy, evictionPolicyGroupVersion)
+	}
+
+	if rs.DeschedulingInterval.Seconds() == 0 {
+		_, err = cron.ParseStandard(rs.DeschedulingIntervalCronExpression)
+		if rs.DeschedulingIntervalCronExpression == "" || err != nil {
+			return fmt.Errorf("Both DeschedulingInterval and deschedulingIntervalCronExpression are not configured. At least one of these two parameter must be configured. ")
+		}
+		klog.V(1).InfoS("Run with deschedulingIntervalCronExpression", "intervalCronExpression", rs.DeschedulingIntervalCronExpression)
+	} else {
+		klog.V(1).InfoS("Run with deschedulingInterval", "interval", rs.DeschedulingInterval)
+	}
+
+	if rs.LeaderElection.LeaderElect && rs.DryRun {
+		klog.V(1).InfoS("Warning: DryRun is set to True. You need to disable it to use Leader Election.")
+	}
+
+	if rs.LeaderElection.LeaderElect && !rs.DryRun {
+		if err := NewLeaderElection(runFn, rsclient, &rs.LeaderElection, ctx); err != nil {
+			return fmt.Errorf("leaderElection: %w", err)
+		}
+		return nil
+	}
+
+	return runFn()
+}
+
+func versionCompatibilityCheck(rs *options.DeschedulerServer) {
+	serverVersion, serverErr := rs.Client.Discovery().ServerVersion()
+	if serverErr != nil {
+		klog.V(1).InfoS("Warning: Get Kubernetes server version fail")
+		return
+	}
+
+	deschedulerMinorVersion := strings.Split(version.Get().Minor, ".")[0]
+	deschedulerMinorVersionFloat, err := strconv.ParseFloat(deschedulerMinorVersion, 64)
+	if err != nil {
+		klog.Warning("Warning: Convert Descheduler minor version to float fail")
+	}
+
+	kubernetesMinorVersionFloat, err := strconv.ParseFloat(serverVersion.Minor, 64)
+	if err != nil {
+		klog.Warning("Warning: Convert Kubernetes server minor version to float fail")
+	}
+
+	if math.Abs(deschedulerMinorVersionFloat-kubernetesMinorVersionFloat) > 3 {
+		klog.Warningf("Warning: Descheduler minor version %v is not supported on your version of Kubernetes %v.%v. See compatibility docs for more info: https://github.com/kubernetes-sigs/descheduler#compatibility-matrix", deschedulerMinorVersion, serverVersion.Major, serverVersion.Minor)
+	}
+}
+
+func cachedClient(
```

Comment:
Why do we need a fakeclient to call here?

Reply:
It's used in `DryRun` mode: )